File: Implementation/Java/Source/surface/model/VCollection.java
package cks.surface.model;

import cks.io.SourcePos;

import java.io.PrintWriter;
import java.util.List;

public class VCollection extends Value
{
	public final List<Element> elements;

	public VCollection(SourcePos loc, List<Element> elements)
	{
		super(loc);
		this.elements = elements;
	}

	protected void dumpChildren(PrintWriter out, int depth)
	{
		for (Element e : elements) {
			e.first.dump(out, depth);
			if (e.arrow != null) {
				indent(out, depth+1);
				out.println("->");
				e.second.dump(out, depth+1);
			}
		}
		super.dumpChildren(out, depth);
	}

	public static final class Element
	{
		public final Value first;
		public final SourcePos arrow; // Might be null.
		public final Value second; // Might be null.

		public Element(Value first, SourcePos arrow, Value second)
		{
			assert first != null;
			assert (arrow == null) == (second == null); // Both must either be null or non-null.
			this.first = first;
			this.arrow = arrow;
			this.second = second;
		}
	}
}

File: Implementation/C#/Source/Text/Reader/Model/VCollection.cs
using System.Collections.Generic;
using StringBuilder = System.Text.StringBuilder;
using TextWriter = System.IO.TextWriter;

namespace Cks.Text.Reader.Model
{

public sealed class VCollection : Value
{
	public readonly List<Element> Elements;

	public VCollection(SourcePos SourcePos, List<Element> Elements) : base(SourcePos)
	{
		this.Elements = Elements;
	}

	public override void ToString(StringBuilder buf)
	{
		buf.Append("[");
		string sep = "";
		foreach (Element e in Elements) {
			buf.Append(sep); sep = ", ";
			e.First.ToString(buf);
			if (e.ArrowPos != null) {
				buf.Append(" -> ");
				e.Second.ToString(buf);
			}
		}
		buf.Append("]");
	}

	protected override string DumpHeader()
	{
		return "[...]";
	}

	protected override void DumpChildren(TextWriter Out, uint Depth)
	{
		foreach (Element e in Elements) {
			e.First.Dump(Out, Depth);
			if (e.ArrowPos != null) {
				Indent(Out, Depth+1);
				Out.WriteLine("->");
				e.Second.Dump(Out, Depth+1);
			}
		}
	}

	public sealed class Element
	{
		public readonly Value First;
		public readonly SourcePos ArrowPos;
		public readonly Value Second;

		public Element(Value First)
		{
			this.First = First;
			this.ArrowPos = null;
			this.Second = null;
		}

		public Element(Value First, SourcePos ArrowPos, Value Second)
		{
			this.First = First;
			this.ArrowPos = ArrowPos;
			this.Second = Second;
		}
	}
}

}
